proxy/http2/Http2ConnectionState.cc

```diff
@@ -645,8 +645,17 @@ rcv_goaway_frame(Http2ConnectionState &cstate, const Http2Frame &frame)
   Http2StreamDebug(cstate.ua_session, stream_id, "GOAWAY: last stream id=%d, error code=%d", goaway.last_streamid,
                    static_cast<int>(goaway.error_code));
 
-  cstate.rx_error_code = {ProxyErrorClass::SSN, static_cast<uint32_t>(goaway.error_code)};
-  cstate.handleEvent(HTTP2_SESSION_EVENT_FINI, nullptr);
+  // If peer indicated an error, go ahead and shutdown hard
+  if (goaway.error_code != Http2ErrorCode::HTTP2_ERROR_NO_ERROR) {
+    cstate.rx_error_code = {ProxyErrorClass::SSN, static_cast<uint32_t>(goaway.error_code)};
+    cstate.send_goaway_frame(cstate.get_latest_stream_id_in(), Http2ErrorCode::HTTP2_ERROR_NO_ERROR);
+    cstate.ua_session->set_half_close_local_flag(true);
+    cstate.handleEvent(HTTP2_SESSION_EVENT_FINI, nullptr);
+  } else {
+    // Otherwise, try to hang around to finish any remaining steams
+    cstate.set_seen_fini();
```

Comment:
1) Can we just schedule `HTTP2_SESSION_EVENT_FINI` instead of these two lines?

IIRC, `fini_received` should be set to `true` only in the process of HTTP2_SESSION_EVENT_FINI event, and the event handler is the one and only place that starts connection closing. We always need to invoke the event handler on either a clean closing or an error closing. See other places that send GOAWAY frame.

2) Don't we need to do the other stuff we do in the error case (in the true branch)?

Don't we need to send GOAWAY this case, or do we send it somewhere else?
Also, I'm not 100% sure but `set_half_close_local_flag(true);` perhaps need to be called in either cases, like other places do.


So, maybe we can differentiate only how we invoke the event (sync or async).

Reply:
The problem with scheduling the HTTP2_SESSION_EVENT_FINI is that the handler processing for that event will immediately cleanup all in process streams, which defeats what we are trying to achieve here.

Reply:
Hmm, I'm confused. We want to return a HTTP response on a stream, but we call release_stream here. What does `release_stream` do here? Is `total_client_streams_count` 0 or 1?

I don't remember what was `fini_received` originally, but it seems like a flag that indicates whether HTTP2_SESSION_EVENT_FINI event is processed. Even if we can achieve the delay with this flag, I'd rather introduce a new flag, `Http2ClientSession::half_close_remote`, and invoke HTTP2_SESSION_EVENT_FINI when all streams finished transfer if `half_close_remote` is true.

Changing the meaning of `fini_received` may break `Http2ConnectionState::is_state_closed`.

Reply:
My interpretation of fini_received flag means that we have seen a FINI frame or otherwise the client side has indicated that the connection is gone.  

The call to release_stream with a nullptr argument just goes through and sees if the state of the world is ready to shutdown (all stream counts set to 0 and a FINI has been seen).

Reply:
> My interpretation of fini_received flag means that we have seen a FINI frame or otherwise the client side has indicated that the connection is gone.

There's no FINI frame.. You mean GOAWAY? Anyways, I don't think your interpretation of the flag is correct. As you can see, `fini_received` flag is set (in the event handler for `HTTP2_SESSION_EVENT_FINI`) regardless of client's status nor action (`send_settings_frame` is a good example).

Also, the event handler for `HTTP2_SESSION_EVENT_FINI` is the only place that calls `SET_HANDLER(&Http2ConnectionState::state_closed);` This suggests that we always need to schedule the event when we close a connection.

So I think the right way to do this is delaying scheduling `HTTP2_SESSION_EVENT_FINI`.